backend/LexBoxApi/Services/ProjectService.cs

```diff
@@ -69,4 +72,39 @@ public async Task<ProjectMigrationStatus> GetProjectMigrationStatus(string proje
         if (migrationStatus == default) throw new NotFoundException($"Project not found: {projectCode}");
         return migrationStatus;
     }
+
+    public async Task MigrateProject(string projectCode)
+    {
+        var project = await _dbContext.Projects.SingleAsync(p => p.Code == projectCode);
+        project.MigrationStatus = ProjectMigrationStatus.Migrating;
+        await _dbContext.SaveChangesAsync();
+        _migrationService.QueueMigration(projectCode);
+    }
+
+    public async Task<bool> AwaitMigration(string projectCode, CancellationToken cancellationToken)
+    {
+        var project = await _dbContext.Projects.SingleAsync(p => p.Code == projectCode, cancellationToken);
+        if (project.MigrationStatus == ProjectMigrationStatus.Migrated) return true;
+        //could be a race condition here... but eh.
+        while (await _migrationService.MigrationCompleted.WaitToReadAsync(cancellationToken) &&
+            _migrationService.MigrationCompleted.TryPeek(out var migratedProjectCode))
+        {
+            if (migratedProjectCode == projectCode)
+            {
+                await _migrationService.MigrationCompleted.ReadAsync(cancellationToken);
+                return true;
+            }
+
+            await Task.Delay(TimeSpan.FromSeconds(5), cancellationToken);
+            //if the project is still in the queue (because no one else is listening) then remove it.
+            // probably won't be true, but just in case.
+            if (_migrationService.MigrationCompleted.TryRead(out var queuedProjectCode) &&
+                queuedProjectCode == projectCode)
+            {
+                return true;
+            }
```

Comment:
I think there's a problem here:
- If 3 projects are being migrated at the same time
- the 1st one complete
- the two remaining "waiters" land at 98
- During the 5s the other 2 migrations complete
- What happens here (lines 101/102) is then a race condition depending on what order the migrations were completed in and when the waiter awake

Assuming I understand how this code actually works, I think you could fix it by replacing 101 - 105 with this:
```
            if (_migrationService.MigrationCompleted.TryPeek(out var queuedProjectCode)
            {
                if (queuedProjectCode == migratedProjectCode) // actually make sure it's the same project that triggered the read on 89 before discarding it
                    await _migrationService.MigrationCompleted.ReadAsync(cancellationToken);
               else if (queuedProjectCode  == projectCode) // I don't think we actually want this case...
               {
                    await _migrationService.MigrationCompleted.ReadAsync(cancellationToken);
                    return true;
               }
            }
```

Reply:
that's a really good point. I agree that's a problem. I'm wondering if there isn't a better way to solve it. Right now the channel is setup so that only 1 reader can get a value. We really want every reader to see every value and then some can just ignore it if they don't care about it. I think we can do that pretty easily with one channel per awaiting call, I'm going to give it a try in another pr and we can see how bad it is, I suspect it'll be simpler but we'll see.